src/engine/renderer/tr_shade.cpp

```diff
@@ -2882,6 +2882,89 @@ void Tess_StageIteratorGeneric()
 	}
 }
 
+
+void Tess_StageIteratorPortal() {
+	int stage;
+
+	// log this call
+	if ( r_logFile->integer ) {
+		// don't just call LogComment, or we will get
+		// a call to va() every frame!
+		GLimp_LogComment( va
+		( "--- Tess_StageIteratorPortal( %s, %i vertices, %i triangles ) ---\n", tess.surfaceShader->name,
+			tess.numVertexes, tess.numIndexes / 3 ) );
+	}
+
+	GL_CheckErrors();
+
+	if ( !glState.currentVBO || !glState.currentIBO || glState.currentVBO == tess.vbo || glState.currentIBO == tess.ibo ) {
+		Tess_UpdateVBOs();
+	}
+
+	// set face culling appropriately
+	if ( backEnd.currentEntity->e.renderfx & RF_SWAPCULL )
+		GL_Cull( ReverseCull( tess.surfaceShader->cullType ) );
+	else
+		GL_Cull( tess.surfaceShader->cullType );
+
+	// set polygon offset if necessary
+	if ( tess.surfaceShader->polygonOffset ) {
+		glEnable( GL_POLYGON_OFFSET_FILL );
+		GL_PolygonOffset( r_offsetFactor->value, r_offsetUnits->value );
+	}
+
+	bool stageFound = false;
+	// call shader function
+	for ( stage = 0; stage < MAX_SHADER_STAGES; stage++ ) {
+		shaderStage_t* pStage = tess.surfaceStages[stage];
+
+		if ( !pStage ) {
+			break;
+		}
+
+		if ( !RB_EvalExpression( &pStage->ifExp, 1.0 ) ) {
+			continue;
+		}
+
+		switch ( pStage->type ) {
+			case stageType_t::ST_COLORMAP:
+			case stageType_t::ST_LIGHTMAP:
+			case stageType_t::ST_DIFFUSEMAP:
+			case stageType_t::ST_COLLAPSE_lighting_PHONG:
+			case stageType_t::ST_COLLAPSE_lighting_PBR:
+			case stageType_t::ST_COLLAPSE_reflection_CB:
+			case stageType_t::ST_REFLECTIONMAP:
+			case stageType_t::ST_REFRACTIONMAP:
+			case stageType_t::ST_DISPERSIONMAP:
+			case stageType_t::ST_SKYBOXMAP:
+			case stageType_t::ST_SCREENMAP:
+			case stageType_t::ST_PORTALMAP:
+			case stageType_t::ST_HEATHAZEMAP:
+			case stageType_t::ST_LIQUIDMAP:
+				Render_generic( pStage );
```

Comment:
What's the motivation of creating a new StageIterator function and why do we want to use Render_generic on things which are usually rendered with a different function?

Reply:
When writing stencil the portal surface should only be rendered once, otherwise it will be rendered with light stages and the like, which would break stencil. (https://github.com/DaemonEngine/Daemon/pull/1031/commits/d0ecf605d3f69f03dac57ef3acce2e1d4a745f19 updates it slightly to make sure only one stage is rendered)
Render_generic seemed to be the most appropriate there since only depth and stencil output are required at that stage, no extra information.

Reply:
The depth only pass can be changed to Tess_StageIteratorDepthFill() though.

Reply:
Actually, even that shouldn't be changed since otherwise it might do a wasteful ST_COLLAPSE_lighting_PBR stage render.